### PR TITLE
Antlr4runtime - upgrade to 4.10

### PR DESCRIPTION
**Pull request by michalwidera**
**Antlr4runtime/4.10**

New version of Antlr4 has appear recently - 10.04.2022. This version has changed major and have bunch of c++ fixes and improvements.
More info here: https://github.com/antlr/antlr4/releases/tag/4.10

---

- [X] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [X] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [X] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [X] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/antlr4-cppruntime/all/conanfile.py [hunk 1]
```diff
@@ -64,18 +75,31 @@ def validate(self):
...
+        else:
+            self.output.warn("{} requires C++17. Your compiler is unknown. Assuming it supports C++17.".format(self.name))
+
+
     @functools.lru_cache(1)
     def _configure_cmake(self):
         cmake = CMake(self)
         cmake.definitions["ANTLR4_INSTALL"] = True
         cmake.definitions["WITH_LIBCXX"] = self.settings.compiler.get_safe("libcxx") == "libc++"
+        cmake.definitions["ANTLR_BUILD_CPP_TESTS"] = "OFF"
```
Comment (SpaceIm) [inline recipes/antlr4-cppruntime/all/conanfile.py:102]:
```suggestion
        cmake.definitions["ANTLR_BUILD_CPP_TESTS"] = False
```

File: recipes/antlr4-cppruntime/all/conanfile.py [hunk 2]
```diff
@@ -64,18 +75,31 @@ def validate(self):
             raise ConanInvalidConfiguration(f"arm architectures are not supported")
             # Need to deal with missing libuuid on Arm.
             # So far ANTLR delivers macOS binary package.
-        compiler, version = self.settings.compiler, tools.Version(self.settings.compiler.version)
-        if compiler == "Visual Studio" and version < "16":
+        compiler, compiler_version = self.settings.compiler, tools.Version(self.settings.compiler.version)
+
+        if compiler == "Visual Studio" and compiler_version < "16":
             raise ConanInvalidConfiguration(f"library claims C2668 'Ambiguous call to overloaded function'")
```
Comment (jgsogo) [inline recipes/antlr4-cppruntime/all/conanfile.py:81]:
```suggestion
            raise ConanInvalidConfiguration("library claims C2668 'Ambiguous call to overloaded function'")
```

File: recipes/antlr4-cppruntime/all/conanfile.py [hunk 3]
```diff
@@ -64,18 +75,31 @@ def validate(self):
             raise ConanInvalidConfiguration(f"arm architectures are not supported")
```
Comment (jgsogo) [inline recipes/antlr4-cppruntime/all/conanfile.py:75]:
```suggestion
            raise ConanInvalidConfiguration("arm architectures are not supported")
```

**Review by ericLemanissier — APPROVED** [review]

**Comment by parrt** [comment]
I just pushed antlr 4.10.1

**Comment by michalwidera** [comment]
**Antlr4runtime/4.10.1**

New version of Antlr4 appear few days after 4.10. Tiny update to fix build issue where java requirement for runtime was 11 not 1.8.
More info here: https://github.com/antlr/antlr4/releases/tag/4.10.1

File: recipes/antlr4-cppruntime/all/conanfile.py [hunk 4]
```diff
@@ -23,7 +23,14 @@ class Antlr4CppRuntimeConan(ConanFile):
         "fPIC": True,
     }
     settings = "os", "compiler", "build_type", "arch"
-    generators = "cmake", "cmake_find_package"
+    generators = "cmake", "pkg_config"
```
Comment (SSE4) [inline recipes/antlr4-cppruntime/all/conanfile.py:26]:
```suggestion
    generators = "cmake", "pkg_config"
    short_paths = True
```

Comment (michalwidera) [reply recipes/antlr4-cppruntime/all/conanfile.py:26]: ok

**Comment by michalwidera** [comment]
Hi - on PR#10287 I'm getting 'unknown' error on build mscv 16 - this is infrastructure issue, no report or logs. How can I fix or diagnose that?

Resp: You patiently wait. The error will be handled by those managing the CI machines.

**Comment by jgsogo** [comment]
@SSE4 , given the latest error, is this an `infrastructure` error?

**Comment by SSE4** [comment]
> @SSE4 , given the latest error, is this an `infrastructure` error?

I don't think so

**Review by jgsogo — COMMENTED** [review]
Existing recipe was working for version `4.9.3`. Unless there is a good reason for it, version `4.9.3` should keep working with previous restrictions (like C++11) and new restrictions should only apply to new versions.

You can use `self.version` (`tools.Version(self.version)`) inside the recipe to apply some checks conditionally.

**Comment by michalwidera** [comment]
gcc - version 5 - for 4.9.3 should be supported. You are right. Going to fix asap.

File: recipes/antlr4-cppruntime/all/conanfile.py [hunk 5]
```diff
@@ -54,28 +63,54 @@ def source(self):
...
+        compiler_version = tools.Version(self.settings.compiler.version)
+        antlr_version = tools.Version(self.version)
+
+        if compiler == "Visual Studio" and compiler_version < "16":
+            raise ConanInvalidConfiguration("library claims C2668 'Ambiguous call to overloaded function'")
             # Compilation of this library on version 15 claims C2668 Error.
             # This could be Bogus error or malformed Antl4 libary.
             # Version 16 compiles this code correctly.
 
+        if antlr_version != "4.9.3":
```
Comment (jgsogo) [inline recipes/antlr4-cppruntime/all/conanfile.py:91]:
```suggestion
        if antlr_version >= "4.10":
```

Probably other `4.9.x` or lower version would work with C++11 as well.

**Review by jgsogo — COMMENTED** [review]
Just one minor comment

File: recipes/antlr4-cppruntime/all/conanfile.py (hunk 5 above)
Comment (michalwidera) [reply recipes/antlr4-cppruntime/all/conanfile.py]: Agree - If someone will try adapt previous versions for some reason that should help.

**Review by jgsogo — APPROVED** [review]

**Review by uilianries — APPROVED** [review]
LGTM

**Review by ericLemanissier — APPROVED** [review]

**Comment by michalwidera** [comment]
Big thanks for all humble reviewers and supporters.

**Pull request merged**